src/IR/LambdaLift.hs

```diff
@@ -0,0 +1,157 @@
+  deriving Functor                      via (StateT LiftCtx Compiler.Pass)
+  deriving Applicative                  via (StateT LiftCtx Compiler.Pass)
+  deriving Monad                        via (StateT LiftCtx Compiler.Pass)
+  deriving MonadFail                    via (StateT LiftCtx Compiler.Pass)
+  deriving (MonadError Compiler.Error)  via (StateT LiftCtx Compiler.Pass)
+  deriving (MonadState LiftCtx)         via (StateT LiftCtx Compiler.Pass)
+
+runLiftFn :: LiftFn a -> Compiler.Pass a
+runLiftFn (LiftFn m) = evalStateT
+  m
+  LiftCtx { globalScope  = S.empty
+          , currentScope = S.empty
+          , currentFrees = S.empty
+          , lifted       = []
+          , anonCount    = 0
+          }
+
+populateGlobalScope :: [(I.VarId, I.Expr Poly.Type)] -> LiftFn ()
+populateGlobalScope defs = do
+  let globalNames = map fst defs
+  modify $ \st -> st { globalScope = S.fromList globalNames }
+
+inCurrentScope :: I.VarId -> LiftFn Bool
+inCurrentScope v = S.member v <$> gets currentScope
+
+addCurrentScope :: I.VarId -> LiftFn ()
+addCurrentScope v =
+  modify $ \st -> st { currentScope = S.insert v $ currentScope st }
+
+getFresh :: LiftFn Int
+getFresh = do
+  curCount <- gets anonCount
+  modify $ \st -> st { anonCount = anonCount st + 1 }
+  return curCount
+
+addLifted :: String -> I.Expr Poly.Type -> LiftFn ()
+addLifted name lam =
+  modify $ \st -> st { lifted = (I.VarId (Identifier name), lam) : lifted st }
+
+addFreeVar :: I.VarId -> LiftFn ()
+addFreeVar v =
+  modify $ \st -> st { currentFrees = S.insert v $ currentFrees st }
+
+newScope :: [I.VarId] -> LiftFn ()
+newScope vs = modify $ \st -> st
+  { currentScope = S.union (globalScope st) (S.fromList vs)
+  , currentFrees = S.empty
+  }
+
+makeLiftedLambda
+  :: [I.Binder] -> I.Expr Poly.Type -> Poly.Type -> LiftFn (I.Expr Poly.Type)
+makeLiftedLambda [] body _ = return body
+makeLiftedLambda vs body t = do
+  liftedBody <- makeLiftedLambda (tail vs) body t
+  traceM (show vs)
+  return (I.Lambda (head vs) liftedBody t)
+
+liftLambdas'
+  :: (I.VarId, I.Expr Poly.Type) -> LiftFn (I.VarId, I.Expr Poly.Type)
+liftLambdas' (v, lam@(I.Lambda _ _ t)) = do
+  let (vs, body) = I.collectLambda lam
+  newScope $ map (\(Just vi) -> vi) vs
+  liftedBody <- liftLambdas body
+  return (v, foldl (\lam' v' -> I.Lambda v' lam' t) liftedBody vs)
+liftLambdas' _ = error "Expected top-level lambda binding"
+
+liftLambdas :: I.Expr Poly.Type -> LiftFn (I.Expr Poly.Type)
+liftLambdas n@(I.Var v _) = do
+  isNotFree <- inCurrentScope v
+  if isNotFree
+    then return n
+    else do
+      addFreeVar v
+      return n
+liftLambdas (I.App e1 e2 t) = do
+  liftedE1 <- liftLambdas e1
+  liftedE2 <- liftLambdas e2
+  return $ I.App liftedE1 liftedE2 t
+liftLambdas (I.Prim p exprs t) = do
+  liftedExprs <- mapM liftLambdas exprs
+  return $ I.Prim p liftedExprs t
+liftLambdas lam@(I.Lambda _ _ t) = do
+  let (vs, body) = I.collectLambda lam
+  traceM "Lambda"
+  oldCtx <- get
```

Comment:
Avoid manually saving and restoring state like this. See the `withNewScope` helper that Xijiao uses in the type checker